src/jemalloc.c

```diff
@@ -1737,6 +1742,24 @@ malloc_init_hard_a0_locked() {
 		return true;
 	}
 	a0 = arena_get(TSDN_NULL, 0, false);
+
+	if (opt_hpa) {
+		/*
+		 * The global HPA uses the edata cache from a0, and so needs to
```

Comment:
Question: should we limit the HPA to auto arenas only? I see that right now it's guarded under custom hooks, but manual arenas with default hooks may still have it? If we want to make that work, arena_reset would need special care for the HPA.

Reply:
I think arena_reset works correctly as is in this implementation -- the slabs / large extents get freed back to PA in the arena-level code, so we'll evict everything from the psset back to the hpa_central before arena destruction.

I'm fairly ambivalent about the broader question of whether or not we ought to (independent of whether or not it's correct to); having manual arenas have drastically different perf characteristics from auto arenas feels wrong, but OTOH most programs currently using manual arenas are probably expecting one set of time/space tradeoffs, and we'd be giving them another.

I think if we're careful with our jemalloc 6 design changes, we can make this question easier to answer.